std/math/ff.asm

```diff
@@ -0,0 +1,60 @@
+/// Inverts `x` in the finite field with modulus `modulus`.
+/// Assumes that `modulus` is prime, but does not check it.
+let inverse = |x, modulus|
+    if x <= 0 || x >= modulus {
+        std::check::panic("Tried to compute the inverse of zero, of a negative number or a number outside the field.")
+    } else {
+        // TODO this is written in a complicated way
+        // because we do not have tuple destructuring assignment
+        (|r|
+            if r[0] < 0 { r[0] + modulus } else { r[0] }
+        )(extended_gcd(x, modulus))
+    };
+
+/// Computes `x + y` modulo the modulus.
+let add = |x, y, modulus| reduce(x + y, modulus);
+
+/// Computes `x - y` modulo the modulus.
+let sub = |x, y, modulus| reduce(x - y, modulus);
+
+/// Computes `x * y` modulo the modulus.
+/// TODO maybe just computing "(x * y) % modulus" is faster...
+let mul = |x, y, modulus|
+    match y {
+        0 => 0,
+        1 => x,
+        _ =>
+            add(
+                if y & 1 == 1 { x } else { 0 },
+                mul(x, y >> 1, modulus) << 1,
+                modulus
+            )
+    };
+
+/// Computes `x / y` modulo the modulus.
+let div = |x, y, modulus| mul(x, inverse(y, modulus), modulus);
+
+let reduce = |x, modulus|
+    if x < 0 {
+        reduce(x + modulus, modulus)
+    } else {
+        if x >= modulus {
+            reduce(x - modulus, modulus)
+        } else {
+            x
+        }
+    };
```

Comment:
```suggestion
let reduce = |x, modulus|
    if x < 0 {
        modulus - (-x % modulus)
    } else {
        x % modulus
    };
```

This should be equivalent and doesn't need iterations.

Reply:
Also, I'm not sure about the name, I first thought it implements the `reduce` function common in functional programming. But either way I think it deserves a comment :)

Reply:
Hmm, I think actually the `x < 0` case should be `(modulus - (-x % modulus)) % modulus` (my suggestion would return `modulus` instead of `0` for `x = -modulus`). A bit complicated, so maybe the original version is better, although it might still do a lot of recursive calls.